File: bpe.java
import java.io.*;
import java.util.*;

public class bpe {
    private static char[] data;

    private static void loadData(String fname) throws IOException {
        File file = new File(fname);
        int arraySize = (int) file.length();

        data = new char[arraySize];
        int index = 0;

        try (FileInputStream fis = new FileInputStream(file)) {
            byte[] buffer = new byte[2000000];
            int bytesRead;

            while ((bytesRead = fis.read(buffer)) != -1) {
                for (int i = 0; i < bytesRead; i++) {
                    data[index++] = (char) (buffer[i] & 0xFF);
                }
            }
        }
    }

    private static class WalkIter {
        private int pos = 0;

        WalkIter() {
            skipHoles();
        }

        private void skipHoles() {
            while (pos < data.length && data[pos] == 0) {
                pos++;
            }
        }

        Integer next() {
            if (pos >= data.length) return null;
            int p = pos;
            pos++;
            skipHoles();
            return p;
        }
    }

    private static class PairWalker {
        private WalkIter walker = new WalkIter();
        private Integer p1 = walker.next();
        private Integer p2 = walker.next();

        IntPair next() {
            Integer a = p1;
            Integer b = p2;
            if (a != null && b != null) {
                p1 = p2;
                p2 = walker.next();
                return new IntPair(a, b);
            }
            return null;
        }

        IntPair afterReplace(int replacePos) {
            p1 = p2;
            p2 = walker.next();
            Integer prevIdx = null;
            if (replacePos > 0) {
                int i = replacePos - 1;
                while (data[i] == 0 && i > 0) { i--; }
                if (data[i] != 0) { prevIdx = i; }
            }
            return new IntPair(prevIdx, p1);
        }
    }

    private static class VPair implements Comparable<VPair> {
        char value1;
        char value2;

        VPair(char value1, char value2) {
            this.value1 = value1;
            this.value2 = value2;
        }

        @Override
        public int compareTo(VPair other) {
            if (value1 != other.value1) {
                return Character.compare(value1, other.value1);
            }
            return Character.compare(value2, other.value2);
        }

        @Override
        public boolean equals(Object o) {
            if (this == o) return true;
            if (o == null || getClass() != o.getClass()) return false;
            VPair vPair = (VPair) o;
            return value1 == vPair.value1 && value2 == vPair.value2;
        }

        @Override
        public int hashCode() {
            return Objects.hash(value1, value2);
        }
    }

    private static Map<VPair, Integer> calcPairHisto() {
        Map<VPair, Integer> h = new HashMap<>();
        PairWalker pairs = new PairWalker();
        IntPair p;
        while ((p = pairs.next()) != null) {
            VPair vp = new VPair(data[p.key], data[p.value]);
            h.put(vp, h.getOrDefault(vp, 0) + 1);
        }
        return h;
    }

    private static record IntPair(Integer key, Integer value) {}
    private static record Triple(VPair first, Integer second, Integer third) {}

    private static Triple mostFreqVal(Map<VPair, Integer> h) {
        VPair minTop = new VPair((char) 0, (char) 0);
        int maxN = 0;
        int total = 0;
        for (Map.Entry<VPair, Integer> entry : h.entrySet()) {
            VPair k = entry.getKey();
            int n = entry.getValue();
            total += n;
            if (n > maxN) {
                maxN = n;
                minTop = k;
            } else if (n == maxN && k.compareTo(minTop) < 0) {
                minTop = k;
            }
        }
        return new Triple(minTop, maxN, total);
    }

    private static void replacePair(VPair oldValuePair, char newVal, Map<VPair, Integer> ph) {
        PairWalker pairs = new PairWalker();
        IntPair indexPair;
        while ((indexPair = pairs.next()) != null) {
            VPair vp = new VPair(data[indexPair.key], data[indexPair.value]);
            if (vp.equals(oldValuePair)) {
                data[indexPair.key] = newVal;
                data[indexPair.value] = 0;
                IntPair afterReplace = pairs.afterReplace(indexPair.key);
                Integer prevIdxOpt = afterReplace.key;
                Integer nextIdxOpt = afterReplace.value;
                if (prevIdxOpt != null) {
                    char leftVal = data[prevIdxOpt];
                    ph.compute(new VPair(leftVal, vp.value1), (k, v) -> (v == null ? 0 : v) - 1);
                    ph.compute(new VPair(leftVal, newVal), (k, v) -> (v == null ? 0 : v) + 1);
                }
                if (nextIdxOpt != null) {
                    char rightVal = data[nextIdxOpt];
                    ph.compute(new VPair(vp.value2, rightVal), (k, v) -> (v == null ? 0 : v) - 1);
                    ph.compute(new VPair(newVal, rightVal), (k, v) -> (v == null ? 0 : v) + 1);
                }
            }
        }
    }

    private static char nextValue = 256;
    private static Map<Character, List<Byte>> thesaurus = new HashMap<>();

    private static boolean oneStep(int step, Map<VPair, Integer> ph) {
        Triple result = mostFreqVal(ph);
        VPair vp = result.first;
        int n = result.second;
        int total = result.third;
        if (n == 1) return true;
        if ((double) total < data.length * 0.707) {
            int newSize = 0;
            for (char c : data) {
                if (c != 0) {
                    data[newSize++] = c;
                }
            }
            data = Arrays.copyOf(data, newSize);
        }
        if (step % 100 == 0) {
            System.out.println("Step " + step + ": n=" + n + " " + (int)vp.value1 + ","
                + (int)vp.value2 + " -> " + (int)nextValue);
        }
        replacePair(vp, nextValue, ph);
        ph.put(vp, 0);
        List<Byte> newThesaurus = new ArrayList<>(thesaurus.get((char) vp.value1));
        newThesaurus.addAll(thesaurus.get((char) vp.value2));
        thesaurus.put(nextValue, newThesaurus);
        nextValue++;
        return false;
    }

    private static void saveTokens(String fname) throws IOException {
        int chunkSize = 1000000;
        String filePath = fname + ".jtok";
        try (RandomAccessFile file = new RandomAccessFile(filePath, "rw")) {
            file.setLength(0);

            char[] buffer = new char[chunkSize];
            int bufferIndex = 0;
            WalkIter w = new WalkIter();

            Map<Character, Integer> h = new HashMap<>();
            Integer pos;
            while ((pos = w.next()) != null) {
                buffer[bufferIndex++] = data[pos];
                h.put(data[pos], h.getOrDefault(data[pos], 0) + 1);
                if (bufferIndex >= chunkSize) {
                    writeBuffer(file, buffer, bufferIndex);
                    bufferIndex = 0;
                }
            }

            if (bufferIndex > 0) {
                writeBuffer(file, buffer, bufferIndex);
            }
        }
    }

    private static void writeBuffer(RandomAccessFile file, char[] buffer, int length) throws IOException {
        byte[] byteBuffer = new byte[length * 2];
        for (int i = 0; i < length; i++) {
            byteBuffer[i * 2] = (byte) buffer[i];
            byteBuffer[i * 2 + 1] = (byte) (buffer[i] >> 8);
        }
        file.write(byteBuffer);
    }

    public static void main(String[] args) throws IOException {
        String fname = args.length > 0 ? args[0] : "enw3";
        for (int x = 0; x <= 255; x++) {
            thesaurus.put((char) x, Collections.singletonList((byte) x));
        }

        loadData(fname);
        System.out.println(data.length);
        Map<VPair, Integer> ph = calcPairHisto();

        for (int i = 1; i <= 65000; i++) {
            if (oneStep(i, ph)) break;
            if (i % 1000 == 0) {
                saveTokens(fname);
            }
        }
        saveTokens(fname);
    }
}

File: bpe.kt
import java.io.File
import java.io.RandomAccessFile
import java.io.FileInputStream
import kotlin.math.min

var data: UShortArray = UShortArray(0)

fun loadData(fname: String) {
    val file = File(fname)
    val arraySize = file.length().toInt()

    data = UShortArray(arraySize)
    var index = 0

    FileInputStream(file).use { fis ->
        val buffer = ByteArray(2000000)
        var bytesRead: Int

        while (fis.read(buffer).also { bytesRead = it } != -1) {
            for (i in 0 until bytesRead) {
                data[index++] = buffer[i].toUByte().toUShort()
            }
        }
    }
}

class WalkIter {
    var pos = 0

    init {
        skipHoles()
    }

    private fun skipHoles() {
        while (pos < data.size && data[pos] == 0.toUShort()) {
            pos++
        }
    }

    fun next(): Int? {
        if (pos >= data.size) return null
        val p = pos
        pos++
        skipHoles()
        return p
    }
}

class PairWalker {
    private var walker = WalkIter()
    private var p1: Int? = walker.next()
    private var p2: Int? = walker.next()

    fun next(): Pair<Int, Int>? {
        val a = p1
        val b = p2
        if (a != null && b != null) {
            p1 = p2
            p2 = walker.next()
            return Pair(a, b)
        }
        return null
    }

    fun afterReplace(replacePos: Int): Pair<Int?, Int?> {
        p1 = p2
        p2 = walker.next()
        var prevIdx: Int? = null
        if (replacePos > 0) {
            var i = replacePos - 1
            while (data[i] == 0.toUShort() && i > 0) { i-- }
            if (data[i] != 0.toUShort()) { prevIdx = i }
        }
        return Pair(prevIdx, p1)
    }
}

data class VPair(val value1: UShort, val value2: UShort) : Comparable<VPair> {
    override fun compareTo(other: VPair): Int {
        return when {
            value1 != other.value1 -> value1.compareTo(other.value1)
            else -> value2.compareTo(other.value2)
        }
    }
}

fun calcPairHisto(): MutableMap<VPair, Int> {
    val h = mutableMapOf<VPair, Int>()
    val pairs = PairWalker()
    while (true) {
        val p = pairs.next() ?: break
        val vp = VPair(data[p.first], data[p.second])
        h[vp] = h.getOrDefault(vp, 0) + 1
    }
    return h
}

fun mostFreqVal(h: MutableMap<VPair, Int>): Triple<VPair, Int, Int> {
    var minTop = VPair(0.toUShort(), 0.toUShort())
    var maxN = 0
    var total = 0
    for ((k, n) in h) {
        total += n
        if (n > maxN) {
            maxN = n
            minTop = k
        } else if (n == maxN && k < minTop) {
            minTop = k
            // minTop = minTop?.let { if (it < k) it else k } ?: k
        }
    }
    return Triple(minTop, maxN, total)
}

fun replacePair(oldValuePair: VPair, newVal: UShort, ph: MutableMap<VPair, Int>) {
    val pairs = PairWalker()
    while (true) {
        val indexPair = pairs.next() ?: break
        val vp = VPair(data[indexPair.first], data[indexPair.second])
        if (vp == oldValuePair) {
            data[indexPair.first] = newVal
            data[indexPair.second] = 0.toUShort()
            val (prevIdxOpt, nextIdxOpt) = pairs.afterReplace(indexPair.first)
            prevIdxOpt?.let { prevIdx ->
                val leftVal = data[prevIdx]
                ph.compute(VPair(leftVal, vp.value1)) { _, v -> (v ?: 0) - 1 }
                ph.compute(VPair(leftVal, newVal)) { _, v -> (v ?: 0) + 1 }
            }
            nextIdxOpt?.let { nextIdx ->
                val rightVal = data[nextIdx]
                ph.compute(VPair(vp.value2, rightVal) ) { _, v -> (v ?: 0) - 1 }
                ph.compute(VPair(newVal, rightVal)) { _, v -> (v ?: 0) + 1 }
            }
        }
    }
}

var nextValue: UShort = 256.toUShort()
var thesaurus: MutableMap<UShort, List<UByte>> = mutableMapOf()

fun oneStep(step: Int, ph: MutableMap<VPair, Int>): Boolean {
    val (vp, n, total) = mostFreqVal(ph)
    if (n == 1) return true
    if (total.toDouble() < data.size * 0.707) {
        var newSize = 0
        for (i in data.indices) {
            if (data[i] != 0.toUShort()) {
                data[newSize++] = data[i]
            }
        }
        data = data.copyOf(newSize)
    }
    if (step % 100 == 0) {
        println("Step $step: n=$n $vp -> $nextValue")
    }
    replacePair(vp, nextValue, ph)
    ph[vp] = 0
    thesaurus[nextValue] = thesaurus[vp.value1]!! + thesaurus[vp.value2]!!
    nextValue++
    return false
}

fun saveTokens(fname: String) {
    val chunkSize = 1000000
    val filePath = "$fname.ktok"
    val file = RandomAccessFile(filePath, "rw")
    file.setLength(0)

    val buffer = UShortArray(chunkSize)
    var bufferIndex = 0
    val w = WalkIter()

    val h = mutableMapOf<UShort, Int>()
    while (true) {
        val pos = w.next() ?: break
        buffer[bufferIndex++] = data[pos]
        h[data[pos]] = h.getOrDefault(data[pos], 0) + 1
        if (bufferIndex >= chunkSize) {
            file.write(buffer.foldIndexed(ByteArray(chunkSize * 2)) { i, acc, us ->
                acc[i * 2] = us.toByte()
                acc[i * 2 + 1] = (us.toInt() shr 8).toByte()
                acc
            })
            bufferIndex = 0
        }
    }

    if (bufferIndex > 0) {
        file.write(buffer.slice(0 until bufferIndex).foldIndexed(ByteArray(bufferIndex * 2)) { i, acc, us ->
            acc[i * 2] = us.toByte()
            acc[i * 2 + 1] = (us.toInt() shr 8).toByte()
            acc
        })
    }

    file.close()
}

fun main(args: Array<String>) {
    val fname = if (args.isNotEmpty()) args[0] else "enw3"
    for (x in 0..255) {
        thesaurus[x.toUShort()] = listOf(x.toUByte())
    }

    loadData(fname)
    println(data.size)
    var ph = calcPairHisto()

    for (i in 1..65000) {
        if (oneStep(i, ph)) break
        if (i % 1000 == 0) {
            saveTokens(fname)
        }
    }
    saveTokens(fname)
}
